lua_modules/mcp23017/mcp23017.lua

```diff
@@ -98,13 +90,68 @@ local function checkPinIsInRange(pin)
     return pin
 end
 
-local function reset(address, i2cId)
-    writeByte(address, i2cId, MCP23017_IODIRA, 0xFF)
-    writeByte(address, i2cId, MCP23017_IODIRB, 0xFF)
+function mcp23017:writeIODIR(bReg, newByte)
+    writeByte(self.address, self.i2cId,
+        bReg and 0x1 --[[const: MCP23017_IODIRB]] or 0x0 --[[const: MCP23017_IODIRA]], newByte)
 end
 
--- setup device
-local function setup(address, i2cId)
+function mcp23017:writeGPIO(bReg, newByte)
+    writeByte(self.address, self.i2cId,
+        bReg and 0x13 --[[const: MCP23017_GPIOB]] or 0x12 --[[const: MCP23017_GPIOA]], newByte)
+end
+
+function mcp23017:readGPIO(bReg)
+    return readByte(self.address, self.i2cId,
+        bReg and 0x13 --[[const: MCP23017_GPIOB]] or 0x12 --[[const: MCP23017_GPIOA]])
+end
+
+-- read pin input
+function mcp23017:getPinState(bReg, pin)
+    return issetBit(readByte(self.address, self.i2cId,
+        bReg and 0x13 --[[const: MCP23017_GPIOB]] or 0x12 --[[const: MCP23017_GPIOA]]),
+        checkPinIsInRange(pin))
+end
+
+-- set pin to low or high
+function mcp23017:setPin(bReg, pin, state)
+    local a, i = self.address, self.i2cId
+    local inReq = bReg and 0x13 --[[const: MCP23017_GPIOB]] or 0x12 --[[const: MCP23017_GPIOA]]
+    local inPin = checkPinIsInRange(pin)
+    local response = readByte(a, i, inReq)
+    writeByte(a, i, inReq,
+        state and setBit(response, inPin) or clearBit(response, inPin))
+    return true
+end
+
+-- set mode for a pin
+function mcp23017:setMode(bReg, pin, mode)
+    local a, i = self.address, self.i2cId
+    local inReq = bReg and 0x1 --[[const: MCP23017_IODIRB]] or 0x0 --[[const: MCP23017_IODIRA]]
+    local inPin = checkPinIsInRange(pin)
+    local response = readByte(a, i, inReq)
+    writeByte(a, i, inReq,
+        mode and setBit(response, inPin) or clearBit(response, inPin))
+    return true
+end
+
+-- reset gpio mode
+function mcp23017:reset()
+    local a, i = self.address, self.i2cId
+    writeByte(a, i, 0x0 --[[const: MCP23017_IODIRA]], 0xFF)
+    writeByte(a, i, 0x1 --[[const: MCP23017_IODIRB]], 0xFF)
+end
+
+-- setup internal pullup
+function mcp23017:setInternalPullUp(bReg, iByte)
+    writeByte(self.address, self.i2cId,
+        bReg and 0x7 --[[const: MCP23017_DEFVALB]] or 0x6 --[[const: MCP23017_DEFVALA]], iByte)
+end
+
+return function(address, i2cId)
+    local self = setmetatable({}, mcp23017)
+
+    self.address = address
+    self.i2cId = i2cId
 
     -- check device address (0x20 to 0x27)
     if (address < 32 or address > 39) then
```

Comment:
Sorry, I think I am misunderstanding your comment.  The version of the code now in `dev` `error`s out in these cases; it did not, AFAICT, "return `nil`" despite appearances: `setup` never returned `false`.  This patch preserves those behaviors (or, well, is at least meant to; did I break something?).

Reply:
>setup never returned false

I missed it completely! Now I wonder if that should be the case -  if the module is not initialised correctly then it shouldn't return the same thing as correctly initialized one. It this a desired behaviour @plomi-net ? I think this can both use `error` and return `nil` but that would be up to discussion.

Reply:
This looks like a mistake. Of course, in case of error nil can be returned. I assumed that throwing an error would be enough.

Reply:
I am inclined to think that throwing an `error` is also enough, yes.